### ISSUE TITLE
reactor.callWhenRunning is not in the Using Processes document

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10835]
|<img alt="hagna's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| hagna reported|
|-|-|
|Trac ID|trac#1246|
|Type|defect|
|Created|2005-10-04 22:50:25Z|
|Branch|https://github.com/twisted/twisted/tree/callwhenrunning-doc-1246|

Attachments:

* [example.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/26c/26ced0b5ea2b6f3c8b1e5090a6cdbdfae5a6f853/05e0372f8165776bed0a53b989eaffd9e3139848.py) (2505 bytes) - added by hagna on 2005-10-04 22:50:24Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1246 1246
type__defect defect
reporter__hagna hagna
priority__normal normal
milestone__ 
branch__branches_callwhenrunning_doc_1246 branches/callwhenrunning-doc-1246
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1128466225000000 1128466225000000
changetime__1220226954000000 1220226954000000
version__ 
owner__ 
cc__hagna cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="hagna's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|hagna commented|
|-|-|

```
#!html
<pre>
The way to get rid of a race condition where spawnprocess is called before the
signal handler is installed is to use reactor.callWhenRunning(spawnprocess ...).
 The examples in "Using Processes" don't mention the race condition or the
solution.  

Here is an example that doesn't use callWhenRunning.  It would work fine if it
was using callWhenRunning.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Shall I create a patch that adds this to the core docs? Can anyone review this example?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

This example is rather complex.  A simpler example would be more like this (untested):

```python
from twisted.internet.utils import getProcessOutput
from twisted.internet import reactor

d = getProcessOutput('/bin/echo', ['/bin/echo', 'foo'])
def cbOutput(output):
    print output
    reactor.stop()
d.addCallback(cbOutput)
reactor.run()
```

This would better be written (for the time being) as

```python
from twisted.internet.utils import getProcessOutput
from twisted.internet import reactor

def getOutput():
    d = getProcessOutput('/bin/echo', ['/bin/echo', 'foo'])
    def cbOutput(output):
        print output
        reactor.stop()
    d.addCallback(cbOutput)
reactor.callWhenRunning(getOutput)
reactor.run()
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

(In [24448]) Branching to 'callwhenrunning-doc-1246'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_5):
> This example is rather complex.  A simpler example would be more like this (untested):
> ...

I added your example in a branch with the terse description hagna offered. Let me know how it looks and fits.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

`spawnprocess` in the prose should be capitalized properly and would probably benefit from being an API link.  The reference to `reactor.callWhenRunning` following it could be an API link too.

I'd suggest dropping the mention of a "signal handler" and just talk about how calling `reactor.spawnProcess` before calling `reactor.run` can lead to a race where process exit isn't noticed (and a warning emitted).  Then it becomes more clear how using `reactor.callWhenRunning` helps.  This should also help make the statement "The way to get rid of a race condition..." more clear, because the actual race condition will have been defined in the document, which it currently isn't.

Also, my example contained a bug - `getProcessOutput` does **not** take argv[0] as part of the argument list, but I included it because I always forget about that inconsistency.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra set owner to @exarkun<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_9):
> `spawnprocess` in the prose should be capitalized properly and would probably benefit from being an API link.  The reference to `reactor.callWhenRunning` following it could be an API link too.
>
> I'd suggest dropping the mention of a "signal handler" and just talk about how calling `reactor.spawnProcess` before calling `reactor.run` can lead to a race where process exit isn't noticed (and a warning emitted).  Then it becomes more clear how using `reactor.callWhenRunning` helps.  This should also help make the statement "The way to get rid of a race condition..." more clear, because the actual race condition will have been defined in the document, which it currently isn't.

Rewrote it in r24642, putting it up for review.

> Also, my example contained a bug - `getProcessOutput` does **not** take argv[0] as part of the argument list, but I included it because I always forget about that inconsistency.

What does this mean, something like this?

```
- d = getProcessOutput('/bin/echo', ['/bin/echo', 'foo'])
+ d = getProcessOutput('/bin/echo')
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

It means `d = getProcessOutput('/bin/echo', ['foo'])`

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_12):
> It means `d = getProcessOutput('/bin/echo', ['foo'])`

Fixed in r24645

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [24652]) Merge callwhenrunning-doc-1246: add a new sample for reactor.callWhenRunning in the Using Processes document.

Author: hagna, exarkun, thijs
Reviewer: exarkun
Fixes: [#10835](https://github.com/twisted/twisted/issues/10835)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
